tsdb/importer/importer.go

```diff
@@ -0,0 +1,529 @@
+
+		blocks[blockIndex] = append(blocks[blockIndex], &sample)
+
+		currentPassCount += 1
+		sampleCount += 1
+		// Have enough samples to write to disk.
+		if currentPassCount == maxSamplesInMemory {
+			if err = flushBlocks(dbDir, blocks, blockMetas, logger); err != nil {
+				return err
+			}
+			// Reset current pass count.
+			currentPassCount = 0
+		}
+	}
+
+	// Flush any remaining samples.
+	if err := flushBlocks(dbDir, blocks, blockMetas, logger); err != nil {
+		return err
+	}
+
+	if err := scanner.Err(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// flushBlocks writes the given blocks of samples to disk, and updates block metadata information.
+func flushBlocks(dbDir string, blocksToFlush [][]*tsdb.MetricSample, blockMetas map[int]*newBlockMeta, logger log.Logger) error {
+	for blockIdx, block := range blocksToFlush {
+		// If current block is empty, nothing to write.
+		if len(block) == 0 {
+			continue
+		}
+		meta := blockMetas[blockIdx]
+		// Put each sample into the appropriate block.
+		bins, binTimes := binSamples(block, meta, tsdb.DefaultBlockDuration)
+		for binIdx, bin := range bins {
+			if len(bin) == 0 {
+				continue
+			}
+			start, end := binTimes[binIdx].start, binTimes[binIdx].end
+			path, err := tsdb.CreateBlock(bin, dbDir, start, end, logger)
+			if err != nil {
+				return err
+			}
+			child := newBlockMeta{
+				index:     blockIdx,
+				count:     0,
+				mint:      start,
+				maxt:      end,
+				isAligned: meta.isAligned,
+				dir:       path,
+			}
+			meta.children = append(meta.children, child)
+		}
+	}
+	return nil
+}
+
+// makeRange returns a series of block times between start and stop,
+// with step divisions.
+func makeRange(start, stop, step int64) []blockTimestampPair {
+	if step <= 0 || stop < start {
+		return []blockTimestampPair{}
+	}
+	r := make([]blockTimestampPair, 0)
+	// In case we only have samples with the same timestamp.
+	if start == stop {
+		r = append(r, blockTimestampPair{
+			start: start,
+			end:   stop + 1,
+		})
+		return r
+	}
+	for s := start; s < stop; s += step {
+		pair := blockTimestampPair{start: s}
+		if (s + step) >= stop {
+			pair.end = stop + 1
+		} else {
+			pair.end = s + step
+		}
+		r = append(r, pair)
+	}
+	return r
+}
+
+// getEmptyBlocks is a simple helper function to create bins for samples.
+func getEmptyBlocks(num int) [][]*tsdb.MetricSample {
+	blocks := make([][]*tsdb.MetricSample, 0)
+	for idx := 0; idx < num; idx++ {
+		block := make([]*tsdb.MetricSample, 0)
+		blocks = append(blocks, block)
+	}
+	return blocks
+}
+
+// mergeBlocks looks for blocks that have overlapping time intervals, and compacts them.
+func mergeBlocks(dbDir string, blockMetas map[int]*newBlockMeta, logger log.Logger) error {
+	for _, blockMeta := range blockMetas {
+		if blockMeta.isAligned {
+			dirs := make([]string, 0)
+			for _, meta := range blockMeta.children {
+				dirs = append(dirs, meta.dir)
+			}
+			path, err := compactBlocks(dbDir, dirs, logger)
```

Comment:
Was experimenting with this (it's awesome, thanks!), and ran into panic errors in compactBlocks() when len(dirs) == 0. Not sure what leads to that case exactly, but for me it happened consistently when running tsdb import on multiple non-overlapping files (i.e. partitioned by date)... first run worked fine then all subsequent runs error out.

Reply:
Thanks for testing this out @clhain! Do you mind explaining the steps you took, and the kind of files you were importing? I think it's a small bug, but I want to be sure before I make a change.

Reply:
Thanks for the quick response! I tried a few things since I thought I was doing something wrong, but the simplest was a single series with 1 sample every 2 hours, in a file per day... so 
e.g.

2019-12-28.txt:
  metric1 {label="whatever"} 1 <timestamp for 2019-12-28 00:00:00>
  metric1 {label="whatever"} 2 <timestamp for 2019-12-28 02:00:00>
  metric1 {label="whatever"} 3 <timestamp for 2019-12-28 04:00:00>
  ...
2019-12-29.txt:
  metric1 {label="whatever"} 1 <timestamp for 2019-12-29 00:00:00>
  ...

Then executing the following:
tsdb import 2019-12-28.txt prometheus-db    <- This worked
tsdb import 2019-12-29.txt prometheus-db   <- This failed (as well as any other file after the first)

I also tried concating the 2 files into a single one, and it worked fine (but all subsequent calls failed).

I wrapped lines 329-333 in a conditional if len(dirs) > 0 and it seems to work... but I'm clueless so wouldn't be surprised if that breaks something else. =)

Here's the output for one of the failures:

level=debug msg="creating temp directory"
level=debug msg="reading existing block metadata to infer time ranges"
level=debug msg="creating new block skeletons based on existing blocks"
level=info msg="importing input data and aligning with existing DB"
level=info msg="write block" mint=1577650456124 maxt=1577656123001 ulid=01DX9TAP0P6TS3JB6WTNJ07FWJ duration=18.699402ms
level=info msg="merging fully overlapping blocks"
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
github.com/prometheus/prometheus/tsdb.compactBlockMetas(0x6a7f2958a5536f01, 0x8213d4d19531d00c, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/github.com/prometheus/prometheus/tsdb/compact.go:335 +0x4b4
github.com/prometheus/prometheus/tsdb.(*LeveledCompactor).Compact(0xc0001562d0, 0xc000134140, 0x16, 0xc0004794f0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x6a7f2958a5536f01, ...)
	/usr/local/go/src/github.com/prometheus/prometheus/tsdb/compact.go:418 +0x6fa
github.com/prometheus/prometheus/tsdb/importer.compactBlocks(0xc000134140, 0x16, 0xc0004794f0, 0x0, 0x0, 0xa6d060, 0xc000132370, 0xc000479478, 0x4cc2c8, 0x0, ...)
	/usr/local/go/src/github.com/prometheus/prometheus/tsdb/importer/importer.go:369 +0x107
github.com/prometheus/prometheus/tsdb/importer.mergeBlocks(0xc000134140, 0x16, 0xc000139590, 0xa6d060, 0xc000132370, 0x0, 0xc0004604e0)
	/usr/local/go/src/github.com/prometheus/prometheus/tsdb/importer/importer.go:329 +0x169
github.com/prometheus/prometheus/tsdb/importer.ImportFromFile(0x7ffd36eaa754, 0x6, 0x7ffd36eaa75b, 0xa, 0x2710, 0xa6d060, 0xc000132370, 0x0, 0x0)
	/usr/local/go/src/github.com/prometheus/prometheus/tsdb/importer/importer.go:109 +0x83c
main.execute(0x0, 0x0)
	/usr/local/go/src/github.com/prometheus/prometheus/tsdb/cmd/tsdb/main.go:146 +0x1efe
main.main()
	/usr/local/go/src/github.com/prometheus/prometheus/tsdb/cmd/tsdb/main.go:45 +0x26
c

Reply:
Thanks for the info! I've changed the relevant section to guard against these kinds of situations, so please give it a shot!